File: services/graph-service/src/main/java/com/chainrisk/graph/logging/StructuredLogger.java
package com.chainrisk.graph.logging;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

/**
 * Structured JSON logger with OpenTelemetry trace correlation.
 */
@Component
public class StructuredLogger {
    
    private static final Logger log = LoggerFactory.getLogger(StructuredLogger.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private final String serviceName;
    
    public StructuredLogger() {
        this.serviceName = "graph-service";
    }
    
    public void debug(String message, Map<String, Object> fields) {
        logStructured("DEBUG", message, fields, null);
    }
    
    public void info(String message, Map<String, Object> fields) {
        logStructured("INFO", message, fields, null);
    }
    
    public void warn(String message, Map<String, Object> fields) {
        logStructured("WARN", message, fields, null);
    }
    
    public void error(String message, Throwable error, Map<String, Object> fields) {
        logStructured("ERROR", message, fields, error);
    }
    
    public void withDuration(String message, long durationMs, Map<String, Object> fields) {
        Map<String, Object> entry = createEntry("INFO", message, fields, null);
        entry.put("duration_ms", durationMs);
        writeLog(entry);
    }
    
    private void logStructured(String level, String message, Map<String, Object> fields, Throwable error) {
        Map<String, Object> entry = createEntry(level, message, fields, error);
        writeLog(entry);
    }
    
    private Map<String, Object> createEntry(String level, String message, 
            Map<String, Object> fields, Throwable error) {
        Map<String, Object> entry = new HashMap<>();
        entry.put("timestamp", Instant.now().toString());
        entry.put("level", level);
        entry.put("service", serviceName);
        entry.put("message", message);
        
        // Add trace context
        Span currentSpan = Span.current();
        SpanContext spanContext = currentSpan.getSpanContext();
        if (spanContext.isValid()) {
            entry.put("trace_id", spanContext.getTraceId());
            entry.put("span_id", spanContext.getSpanId());
        }
        
        // Add custom fields
        if (fields != null && !fields.isEmpty()) {
            entry.put("fields", fields);
        }
        
        // Add error details
        if (error != null) {
            entry.put("error", formatError(error));
        }
        
        return entry;
    }
    
    private String formatError(Throwable error) {
        StringBuilder sb = new StringBuilder();
        sb.append(error.getClass().getName()).append(": ").append(error.getMessage());
        
        // Add first few stack frames
        StackTraceElement[] stack = error.getStackTrace();
        int framesToShow = Math.min(5, stack.length);
        for (int i = 0; i < framesToShow; i++) {
            sb.append("\n\tat ").append(stack[i]);
        }
        
        return sb.toString();
    }
    
    private void writeLog(Map<String, Object> entry) {
        try {
            String json = objectMapper.writeValueAsString(entry);
            // Use standard logger which will be captured by Promtail
            String level = (String) entry.get("level");
            switch (level) {
                case "DEBUG":
                    log.debug(json);
                    break;
                case "WARN":
                    log.warn(json);
                    break;
                case "ERROR":
                    log.error(json);
                    break;
                default:
                    log.info(json);
            }
        } catch (JsonProcessingException e) {
            log.error("Failed to serialize log entry", e);
        }
    }
    
    /**
     * Create a builder for fluent logging.
     */
    public LogBuilder builder(String message) {
        return new LogBuilder(this, message);
    }
    
    public static class LogBuilder {
        private final StructuredLogger logger;
        private final String message;
        private final Map<String, Object> fields = new HashMap<>();
        private Long durationMs;
        private Throwable error;
        
        LogBuilder(StructuredLogger logger, String message) {
            this.logger = logger;
            this.message = message;
        }
        
        public LogBuilder field(String key, Object value) {
            fields.put(key, value);
            return this;
        }
        
        public LogBuilder duration(long ms) {
            this.durationMs = ms;
            return this;
        }
        
        public LogBuilder error(Throwable e) {
            this.error = e;
            return this;
        }
        
        public void debug() {
            logger.debug(message, fields);
        }
        
        public void info() {
            if (durationMs != null) {
                logger.withDuration(message, durationMs, fields);
            } else {
                logger.info(message, fields);
            }
        }
        
        public void warn() {
            logger.warn(message, fields);
        }
        
        public void error() {
            logger.error(message, error, fields);
        }
    }
}

File: services/bff/src/common/structured-logger.ts
import { Injectable } from "@nestjs/common";
import { trace, SpanContext } from "@opentelemetry/api";

export interface LogEntry {
  timestamp: string;
  level: string;
  service: string;
  trace_id?: string;
  span_id?: string;
  message: string;
  duration_ms?: number;
  error?: string;
  fields?: Record<string, unknown>;
}

/**
 * Structured JSON logger with OpenTelemetry trace correlation.
 */
@Injectable()
export class StructuredLogger {
  private serviceName: string;

  constructor(serviceName: string = "bff") {
    this.serviceName = serviceName;
  }

  private createEntry(
    level: string,
    message: string,
    fields?: Record<string, unknown>,
  ): LogEntry {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      service: this.serviceName,
      message,
    };

    // Extract trace context from current span
    const span = trace.getActiveSpan();
    if (span) {
      const ctx: SpanContext = span.spanContext();
      entry.trace_id = ctx.traceId;
      entry.span_id = ctx.spanId;
    }

    if (fields && Object.keys(fields).length > 0) {
      entry.fields = fields;
    }

    return entry;
  }

  private log(entry: LogEntry): void {
    console.log(JSON.stringify(entry));
  }

  debug(message: string, fields?: Record<string, unknown>): void {
    this.log(this.createEntry("DEBUG", message, fields));
  }

  info(message: string, fields?: Record<string, unknown>): void {
    this.log(this.createEntry("INFO", message, fields));
  }

  warn(message: string, fields?: Record<string, unknown>): void {
    this.log(this.createEntry("WARN", message, fields));
  }

  error(
    message: string,
    error?: Error,
    fields?: Record<string, unknown>,
  ): void {
    const entry = this.createEntry("ERROR", message, fields);
    if (error) {
      entry.error = error.stack || error.message;
    }
    this.log(entry);
  }

  withDuration(
    message: string,
    durationMs: number,
    fields?: Record<string, unknown>,
  ): void {
    const entry = this.createEntry("INFO", message, fields);
    entry.duration_ms = durationMs;
    this.log(entry);
  }

  /**
   * Create a child logger with additional default fields.
   */
  child(defaultFields: Record<string, unknown>): ChildLogger {
    return new ChildLogger(this, defaultFields);
  }
}

class ChildLogger {
  constructor(
    private parent: StructuredLogger,
    private defaultFields: Record<string, unknown>,
  ) {}

  private mergeFields(fields?: Record<string, unknown>): Record<string, unknown> {
    return { ...this.defaultFields, ...fields };
  }

  debug(message: string, fields?: Record<string, unknown>): void {
    this.parent.debug(message, this.mergeFields(fields));
  }

  info(message: string, fields?: Record<string, unknown>): void {
    this.parent.info(message, this.mergeFields(fields));
  }

  warn(message: string, fields?: Record<string, unknown>): void {
    this.parent.warn(message, this.mergeFields(fields));
  }

  error(
    message: string,
    error?: Error,
    fields?: Record<string, unknown>,
  ): void {
    this.parent.error(message, error, this.mergeFields(fields));
  }
}

// Default logger instance
export const logger = new StructuredLogger("bff");

/**
 * Timing decorator for methods.
 */
export function Timed(operationName?: string) {
  return function (
    target: unknown,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: unknown[]) {
      const start = Date.now();
      const name = operationName || propertyKey;

      try {
        const result = await originalMethod.apply(this, args);
        const duration = Date.now() - start;
        logger.withDuration(`${name} completed`, duration, { success: true });
        return result;
      } catch (error) {
        const duration = Date.now() - start;
        logger.error(`${name} failed`, error as Error, { duration_ms: duration });
        throw error;
      }
    };

    return descriptor;
  };
}
